File: scripts/load_versification.py
"""Load versification schemes, book identity and the WLC-KJV verse map.

Three tables, all of them data that already existed somewhere and was being
re-derived, guessed at, or silently lost:

* `core.editions_versification` — which tradition each edition numbers by.
* `core.edition_books` — `(edition, article code) -> OSIS book id`, because LHB
  and ESV disagree about four Old Testament book codes and a join on the raw
  code returns silence rather than an error.
* `core.verse_map` — the 1,978 mappings in `wlc/VerseMap.xml`, which the WLC
  ingest read past and discarded.

Nothing is written until it validates. `--validate-only` runs every check and
writes nothing; `--dry-run` prints what would be written. The checks are the
point of the script, so they run on a real load too, and a failure refuses the
whole load rather than leaving a half-populated map.

Usage (from the repository root):
    uv run python scripts/load_versification.py --validate-only
    uv run python scripts/load_versification.py
"""

from __future__ import annotations

import argparse
import asyncio
import sys
import xml.etree.ElementTree as ET
from collections import defaultdict
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from ingest_wlc import LHB_TO_OSIS  # noqa: E402

VERSEMAP_NS = "{http://www.APTBibleTools.com/namespace}"

#: LHB and WLC are laid out one-for-one and share LHB's article codes; ESV
#: differs on exactly four Old Testament books and adds the New Testament.
#: Recorded as an override on `LHB_TO_OSIS` rather than as a second full table,
#: so the four that actually differ are visible instead of buried in 39 rows.
ESV_OT_OVERRIDES = {"EC": "Eccl", "HOS": "Hos", "MI": "Mic", "NA": "Nah"}
ESV_OT_DROPPED = {"ECC", "HO", "MIC", "NAH"}

ESV_NT = {
    "MT": "Matt", "MK": "Mark", "LU": "Luke", "JN": "John", "AC": "Acts",
    "RO": "Rom", "1CO": "1Cor", "2CO": "2Cor", "GA": "Gal", "EPH": "Eph",
    "PHP": "Phil", "COL": "Col", "1TH": "1Thess", "2TH": "2Thess",
    "1TI": "1Tim", "2TI": "2Tim", "TIT": "Titus", "PHM": "Phlm", "HEB": "Heb",
    "JAM": "Jas", "1PE": "1Pet", "2PE": "2Pet", "1JN": "1John", "2JN": "2John",
    "3JN": "3John", "JUD": "Jude", "REV": "Rev",
}

#: Which tradition each edition numbers by. LHB and WLC are Hebrew editions and
#: number as the Masoretic text does; ESV follows the English tradition, which
#: is what `VerseMap.xml` calls KJV.
SCHEMES = {
    "LHB": ("hebrew", "Lexham Hebrew Bible; Masoretic verse numbering"),
    "WLC": ("hebrew", "Westminster Leningrad Codex; Masoretic verse numbering"),
    "ESV": ("english", "English Standard Version; KJV-tradition verse numbering"),
}


def edition_book_rows() -> list[dict]:
    """One row per `(edition, code)` the corpus actually addresses a book by."""
    rows = []
    # `LHB_TO_OSIS` is written in canonical order, so enumerate gives the
    # ordinal without a second list to keep in step with the first.
    for edition in ("LHB", "WLC"):
        for ordinal, (code, osis) in enumerate(LHB_TO_OSIS.items(), start=1):
            rows.append(
                {"edition_key": edition, "code": code, "osis_id": osis, "ordinal": ordinal}
            )

    esv = {c: o for c, o in LHB_TO_OSIS.items() if c not in ESV_OT_DROPPED}
    esv.update(ESV_OT_OVERRIDES)
    esv.update(ESV_NT)
    for ordinal, (code, osis) in enumerate(esv.items(), start=1):
        rows.append(
            {"edition_key": "ESV", "code": code, "osis_id": osis, "ordinal": ordinal}
        )
    return rows


def split_ref(raw: str) -> tuple[str, str | None]:
    """Split `"Isa.63.19!b"` into its reference and its half-verse part.

    The `!a`/`!b` suffix is how the source writes a verse that begins partway
    through another. It is kept as its own column rather than folded into the
    reference, because a reader that does not understand parts must be able to
    see that this row is not a whole-verse equivalence.
    """
    if "!" in raw:
        ref, part = raw.split("!", 1)
        return ref, part
    return raw, None


def parse_verse_map(path: Path, source: str) -> list[dict]:
    """Every mapping in `VerseMap.xml`, as rows.

    `wlc=` is the Hebrew-scheme reference and `kjv=` the English one. The file
    names its second side KJV; ESV follows the same tradition, which the
    validation below checks against the corpus rather than assuming.
    """
    root = ET.parse(path).getroot()
    rows = []
    for book in root.findall(f"{VERSEMAP_NS}book"):
        for verse in book.findall(f"{VERSEMAP_NS}verse"):
            from_ref, from_part = split_ref(verse.get("wlc") or "")
            to_ref, to_part = split_ref(verse.get("kjv") or "")
            mapping_type = verse.get("type") or "full"
            rows.append(
                {
                    "from_scheme": "hebrew",
                    "to_scheme": "english",
                    "from_ref": from_ref,
                    "to_ref": to_ref,
                    "from_part": from_part,
                    "to_part": to_part,
                    "mapping_type": mapping_type,
                    "source": source,
                }
            )
    return rows


def check_map_shape(rows: list[dict]) -> list[str]:
    """Structural checks on the map, independent of any corpus."""
    problems = []
    partials = [r for r in rows if r["mapping_type"] == "partial"]
    parted = [r for r in rows if r["from_part"] or r["to_part"]]

    for row in rows:
        if row["mapping_type"] not in ("full", "partial"):
            problems.append(f"unknown mapping type {row['mapping_type']!r}")
        if row["mapping_type"] == "full" and (row["from_part"] or row["to_part"]):
            problems.append(f"a 'full' mapping carries a part: {row['from_ref']}")
        for side in ("from_ref", "to_ref"):
            if len(row[side].split(".")) != 3:
                problems.append(f"{side} is not book.chapter.verse: {row[side]!r}")

    # Every partial should be marked as one. If a row carried an `!a` without
    # `type="partial"` the check constraint would reject the load; catching it
    # here says which row rather than which constraint.
    if {id(r) for r in partials} != {id(r) for r in parted}:
        problems.append(
            f"{len(partials)} rows typed partial but {len(parted)} carry a part"
        )

    seen = defaultdict(list)
    for row in rows:
        seen[(row["from_ref"], row["from_part"])].append(row["to_ref"])
    for key, targets in seen.items():
        if len(targets) > 1:
            problems.append(f"{key} maps to several references: {targets}")
    return problems


async def validate_against_corpus(engine, rows: list[dict], books: list[dict]) -> list[str]:
    """Check the map verse-by-verse against the corpus it will be used on.

    The map is WLC-to-KJV; this corpus holds ESV. Book- and chapter-level
    agreement had been checked before this script existed, verse-by-verse had
    not, and "the ESV follows the KJV tradition" is an assumption worth
    measuring rather than repeating.
    """
    import sqlalchemy as sa

    problems = []
    osis_of = {(b["edition_key"], b["code"]): b["osis_id"] for b in books}

    sql = sa.text(
        "SELECT d.metadata->>'edition_key' AS ed, "
        "       split_part(d.metadata->>'article', '.', 2) AS code, "
        "       (n.metadata->>'chapter')::int AS chapter, "
        "       (n.metadata->>'verse')::int AS verse "
        "FROM core.documents d "
        "JOIN core.document_nodes n ON n.document_id = d.id "
        "WHERE n.node_type = 'verse' "
        "  AND d.metadata->>'edition_key' IN ('LHB', 'WLC', 'ESV')"
    )
    async with engine.connect() as conn:
        rows_db = (await conn.execute(sql)).fetchall()

    present: dict[str, set[str]] = defaultdict(set)
    for edition, code, chapter, verse in rows_db:
        osis = osis_of.get((edition, code))
        if osis is None:
            problems.append(f"{edition} book code {code!r} has no OSIS id")
            continue
        present[edition].add(f"{osis}.{chapter}.{verse}")

    # 1. Every Hebrew side of the map must exist in both Hebrew editions.
    for edition in ("WLC", "LHB"):
        missing = sorted(
            {r["from_ref"] for r in rows} - present[edition]
        )
        if missing:
            problems.append(
                f"{len(missing)} mapped Hebrew references absent from {edition}: "
                f"{missing[:5]}"
            )

    # 2. Every English side must exist in ESV. This is the assumption under
    #    test: a KJV map is only usable here if its targets are ESV verses.
    missing_esv = sorted({r["to_ref"] for r in rows} - present["ESV"])
    if missing_esv:
        problems.append(
            f"{len(missing_esv)} mapped English references absent from ESV: "
            f"{missing_esv[:8]}"
        )

    # 3. The completeness check the brief names: repairing book identity must
    #    surface exactly the 27 books VerseMap knows about.
    mapped_books = {r["from_ref"].split(".")[0] for r in rows}
    divergent = set()
    heb_by_chapter: dict[tuple[str, int], int] = defaultdict(int)
    eng_by_chapter: dict[tuple[str, int], int] = defaultdict(int)
    for ref in present["LHB"]:
        book, chapter, _ = ref.split(".")
        heb_by_chapter[(book, int(chapter))] += 1
    for ref in present["ESV"]:
        book, chapter, _ = ref.split(".")
        eng_by_chapter[(book, int(chapter))] += 1
    for key, count in heb_by_chapter.items():
        if eng_by_chapter.get(key, -1) != count:
            divergent.add(key[0])
    if divergent != mapped_books:
        problems.append(
            f"divergent books {sorted(divergent)} != VerseMap books "
            f"{sorted(mapped_books)}"
        )
    return problems


async def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--verse-map", type=Path, default=Path("/tmp/morphhb/wlc/VerseMap.xml"))
    ap.add_argument("--source", default=None, help="provenance string for the rows")
    ap.add_argument("--dry-run", action="store_true")
    ap.add_argument("--validate-only", action="store_true")
    args = ap.parse_args()

    import sqlalchemy as sa

    from research_engine.adapters.storage.postgres.schema import (
        edition_books,
        editions_versification,
        verse_map,
    )
    from research_engine.composition import build_container
    from research_engine.config import load_settings

    source = args.source
    if source is None:
        from ingest_wlc import morphhb_commit

        commit = morphhb_commit(args.verse_map.parent.parent)
        source = f"morphhb@{commit[:7]} wlc/VerseMap.xml"

    books = edition_book_rows()
    mappings = parse_verse_map(args.verse_map, source)
    print(f"{len(books)} book codes, {len(mappings)} verse mappings from {source}")
    partials = [m for m in mappings if m["mapping_type"] == "partial"]
    print(
        f"  {len(mappings) - len(partials)} full, {len(partials)} partial, "
        f"{len({m['from_ref'].split('.')[0] for m in mappings})} books"
    )

    problems = check_map_shape(mappings)
    if problems:
        print("REFUSED — the map does not validate:")
        for p in problems[:10]:
            print(f"  {p}")
        return 1

    container = await build_container(load_settings())
    try:
        corpus_problems = await validate_against_corpus(container.engine, mappings, books)
        if corpus_problems:
            print("REFUSED — the map does not agree with the corpus:")
            for p in corpus_problems[:10]:
                print(f"  {p}")
            return 1
        print("validated: shape and corpus agreement both clean")

        if args.validate_only or args.dry_run:
            print("nothing written")
            return 0

        scheme_rows = [
            {"edition_key": key, "scheme": scheme, "notes": notes}
            for key, (scheme, notes) in SCHEMES.items()
        ]
        # Every edition in `bibliography.editions` must get a scheme, so an
        # edition added since this script was written fails loudly here rather
        # than silently defaulting to somebody's tradition.
        async with container.engine.connect() as conn:
            known = {
                row[0]
                for row in (
                    await conn.execute(sa.text("SELECT edition_key FROM bibliography.editions"))
                ).fetchall()
            }
        unscheduled = known - set(SCHEMES)
        if unscheduled:
            print(f"REFUSED — editions with no scheme declared: {sorted(unscheduled)}")
            return 1

        async with container.engine.begin() as conn:
            await conn.execute(sa.delete(verse_map).where(verse_map.c.source == source))
            await conn.execute(sa.delete(edition_books))
            await conn.execute(sa.delete(editions_versification))
            await conn.execute(editions_versification.insert(), scheme_rows)
            await conn.execute(edition_books.insert(), books)
            await conn.execute(verse_map.insert(), mappings)

        print(
            f"wrote {len(scheme_rows)} schemes, {len(books)} book codes, "
            f"{len(mappings)} mappings"
        )
        return 0
    finally:
        await container.close()


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: crates/marginalia-parse/src/bin/load_versification.rs
//! `scripts/load_versification.py` (pure parts): versification tables.
//!
//! The database writes stay Python; what ports is everything checkable
//! without one: the edition book rows, the verse-map parse, the shape
//! checks, and the corpus-agreement comparison over caller-supplied verse
//! sets. Prints counts and problems, mirroring the script's own output,
//! plus `--dump-rows` for the machine-readable tables.

use std::collections::{HashMap, HashSet};

use marginalia_parse::xml::{parse_xml, TextModel};
use serde_json::{json, Value};

const VERSEMAP_NS: &str = "http://www.APTBibleTools.com/namespace";

/// LHB's Logos book codes to morphhb's OSIS filenames, in canonical order
/// (duplicated from `ingest_wlc`, whose table this is; enumerate gives the
/// ordinal so no second list drifts out of step with the first).
const LHB_TO_OSIS: [(&str, &str); 39] = [
    ("GE", "Gen"),
    ("EX", "Exod"),
    ("LE", "Lev"),
    ("NU", "Num"),
    ("DE", "Deut"),
    ("JOS", "Josh"),
    ("JDG", "Judg"),
    ("RU", "Ruth"),
    ("1SA", "1Sam"),
    ("2SA", "2Sam"),
    ("1KI", "1Kgs"),
    ("2KI", "2Kgs"),
    ("1CH", "1Chr"),
    ("2CH", "2Chr"),
    ("EZR", "Ezra"),
    ("NE", "Neh"),
    ("ES", "Esth"),
    ("JOB", "Job"),
    ("PS", "Ps"),
    ("PR", "Prov"),
    ("ECC", "Eccl"),
    ("SO", "Song"),
    ("IS", "Isa"),
    ("JE", "Jer"),
    ("LA", "Lam"),
    ("EZE", "Ezek"),
    ("DA", "Dan"),
    ("HO", "Hos"),
    ("JOE", "Joel"),
    ("AM", "Amos"),
    ("OB", "Obad"),
    ("JON", "Jonah"),
    ("MIC", "Mic"),
    ("NAH", "Nah"),
    ("HAB", "Hab"),
    ("ZEP", "Zeph"),
    ("HAG", "Hag"),
    ("ZEC", "Zech"),
    ("MAL", "Mal"),
];

/// ESV differs on exactly four Old Testament books: an override on the LHB
/// table rather than a second full table, so the four stay visible.
const ESV_OT_OVERRIDES: [(&str, &str); 4] =
    [("EC", "Eccl"), ("HOS", "Hos"), ("MI", "Mic"), ("NA", "Nah")];
const ESV_OT_DROPPED: [&str; 4] = ["ECC", "HO", "MIC", "NAH"];

const ESV_NT: [(&str, &str); 27] = [
    ("MT", "Matt"),
    ("MK", "Mark"),
    ("LU", "Luke"),
    ("JN", "John"),
    ("AC", "Acts"),
    ("RO", "Rom"),
    ("1CO", "1Cor"),
    ("2CO", "2Cor"),
    ("GA", "Gal"),
    ("EPH", "Eph"),
    ("PHP", "Phil"),
    ("COL", "Col"),
    ("1TH", "1Thess"),
    ("2TH", "2Thess"),
    ("1TI", "1Tim"),
    ("2TI", "2Tim"),
    ("TIT", "Titus"),
    ("PHM", "Phlm"),
    ("HEB", "Heb"),
    ("JAM", "Jas"),
    ("1PE", "1Pet"),
    ("2PE", "2Pet"),
    ("1JN", "1John"),
    ("2JN", "2John"),
    ("3JN", "3John"),
    ("JUD", "Jude"),
    ("REV", "Rev"),
];

/// One `(edition, code)` the corpus actually addresses a book by.
#[derive(Debug, Clone)]
struct BookRow {
    edition_key: String,
    code: String,
    osis_id: String,
    ordinal: u64,
}

fn edition_book_rows() -> Vec<BookRow> {
    let mut rows = Vec::new();
    for edition in ["LHB", "WLC"] {
        for (ordinal, (code, osis)) in LHB_TO_OSIS.iter().enumerate() {
            rows.push(BookRow {
                edition_key: edition.to_owned(),
                code: code.to_string(),
                osis_id: osis.to_string(),
                ordinal: ordinal as u64 + 1,
            });
        }
    }
    let mut esv: Vec<(&str, &str)> = LHB_TO_OSIS
        .iter()
        .filter(|(code, _)| !ESV_OT_DROPPED.contains(code))
        .copied()
        .collect();
    esv.extend(ESV_OT_OVERRIDES);
    esv.extend(ESV_NT);
    for (ordinal, (code, osis)) in esv.iter().enumerate() {
        rows.push(BookRow {
            edition_key: "ESV".to_owned(),
            code: code.to_string(),
            osis_id: osis.to_string(),
            ordinal: ordinal as u64 + 1,
        });
    }
    rows
}

/// Python `repr` of a string: single quotes, backslash escapes. The script
/// formats every problem with `{...!r}`, so byte-identical messages need it.
fn py_repr(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('\'');
    for c in s.chars() {
        match c {
            '\'' => out.push_str("\\'"),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 || (c as u32) == 0x7f => {
                out.push_str(&format!("\\x{:02x}", c as u32));
            }
            c => out.push(c),
        }
    }
    out.push('\'');
    out
}

/// Python `repr` of a string list: `['a', 'b']`.
fn py_list(items: &[&str]) -> String {
    format!(
        "[{}]",
        items
            .iter()
            .map(|s| py_repr(s))
            .collect::<Vec<_>>()
            .join(", ")
    )
}

/// Split `"Isa.63.19!b"` into its reference and its half-verse part: the
/// `!a`/`!b` suffix is how the source writes a verse beginning partway
/// through another, kept as its own column so a reader that does not
/// understand parts still sees this row is not a whole-verse equivalence.
fn split_ref(raw: &str) -> (String, Option<String>) {
    match raw.split_once('!') {
        Some((reference, part)) => (reference.to_owned(), Some(part.to_owned())),
        None => (raw.to_owned(), None),
    }
}

#[derive(Debug, Clone)]
struct MapRow {
    from_ref: String,
    to_ref: String,
    from_part: Option<String>,
    to_part: Option<String>,
    mapping_type: String,
    source: String,
}

/// Every mapping in `VerseMap.xml`, as rows. `wlc=` is the Hebrew-scheme
/// reference and `kjv=` the English one.
fn parse_verse_map(raw: &[u8], source: &str) -> Result<Vec<MapRow>, String> {
    let document = parse_xml(raw, TextModel::Et).map_err(|err| err.to_string())?;
    let mut rows = Vec::new();
    for book in document
        .root
        .child_elements()
        .filter(|el| el.tag() == format!("{{{VERSEMAP_NS}}}book"))
    {
        for verse in book
            .child_elements()
            .filter(|el| el.tag() == format!("{{{VERSEMAP_NS}}}verse"))
        {
            let (from_ref, from_part) = split_ref(verse.attr("wlc").unwrap_or(""));
            let (to_ref, to_part) = split_ref(verse.attr("kjv").unwrap_or(""));
            rows.push(MapRow {
                from_ref,
                to_ref,
                from_part,
                to_part,
                mapping_type: verse.attr("type").unwrap_or("full").to_owned(),
                source: source.to_owned(),
            });
        }
    }
    Ok(rows)
}

/// Structural checks on the map, independent of any corpus.
fn check_map_shape(rows: &[MapRow]) -> Vec<String> {
    let mut problems = Vec::new();
    for row in rows {
        if row.mapping_type != "full" && row.mapping_type != "partial" {
            problems.push(format!(
                "unknown mapping type {}",
                py_repr(&row.mapping_type)
            ));
        }
        if row.mapping_type == "full" && (row.from_part.is_some() || row.to_part.is_some()) {
            problems.push(format!("a 'full' mapping carries a part: {}", row.from_ref));
        }
        for side in [&row.from_ref, &row.to_ref] {
            if side.split('.').count() != 3 {
                problems.push(format!(
                    "reference is not book.chapter.verse: {}",
                    py_repr(side)
                ));
            }
        }
    }

    // Every partial is marked as one, and every part-carrier is typed one:
    // the row objects are distinct, so index sets stand in for `id()` sets.
    let partials: HashSet<usize> = rows
        .iter()
        .enumerate()
        .filter(|(_, row)| row.mapping_type == "partial")
        .map(|(i, _)| i)
        .collect();
    let parted: HashSet<usize> = rows
        .iter()
        .enumerate()
        .filter(|(_, row)| row.from_part.is_some() || row.to_part.is_some())
        .map(|(i, _)| i)
        .collect();
    if partials != parted {
        problems.push(format!(
            "{} rows typed partial but {} carry a part",
            partials.len(),
            parted.len()
        ));
    }

    let mut seen: HashMap<(&str, Option<&str>), Vec<&str>> = HashMap::new();
    for row in rows {
        seen.entry((row.from_ref.as_str(), row.from_part.as_deref()))
            .or_default()
            .push(row.to_ref.as_str());
    }
    for ((from_ref, from_part), targets) in &seen {
        if targets.len() > 1 {
            let key = match from_part {
                Some(part) => format!("({}, {})", py_repr(from_ref), py_repr(part)),
                None => format!("({}, None)", py_repr(from_ref)),
            };
            problems.push(format!(
                "{key} maps to several references: {}",
                py_list(targets)
            ));
        }
    }
    problems
}

/// Check the map verse-by-verse against caller-supplied verse rows: the map
/// is WLC-to-KJV, the corpus holds ESV, and "the ESV follows the KJV
/// tradition" is the assumption under test. Rows arrive as the database
/// holds them — `(code, chapter, verse)` per edition — and resolve to OSIS
/// references through the book table, unresolvable codes failing loudly.
fn validate_against_corpus(
    rows: &[MapRow],
    books: &[BookRow],
    raw: &HashMap<String, Vec<(String, u64, u64)>>,
) -> Vec<String> {
    let mut problems = Vec::new();
    let osis_of: HashMap<(&str, &str), &str> = books
        .iter()
        .map(|b| {
            (
                (b.edition_key.as_str(), b.code.as_str()),
                b.osis_id.as_str(),
            )
        })
        .collect();
    let mut present: HashMap<&str, HashSet<String>> = HashMap::new();
    for (edition, verses) in raw {
        for (code, chapter, verse) in verses {
            match osis_of.get(&(edition.as_str(), code.as_str())) {
                Some(osis) => {
                    present
                        .entry(edition.as_str())
                        .or_default()
                        .insert(format!("{osis}.{chapter}.{verse}"));
                }
                None => problems.push(format!(
                    "{edition} book code {} has no OSIS id",
                    py_repr(code)
                )),
            }
        }
    }

    // 1. Every Hebrew side must exist in both Hebrew editions.
    for edition in ["WLC", "LHB"] {
        let empty = HashSet::new();
        let have = present.get(edition).unwrap_or(&empty);
        let mut missing: Vec<&str> = rows
            .iter()
            .map(|r| r.from_ref.as_str())
            .filter(|r| !have.contains(*r))
            .collect::<HashSet<_>>()
            .into_iter()
            .collect();
        missing.sort_unstable();
        if !missing.is_empty() {
            let head: Vec<&str> = missing.iter().take(5).copied().collect();
            problems.push(format!(
                "{} mapped Hebrew references absent from {edition}: {}",
                missing.len(),
                py_list(&head)
            ));
        }
    }

    // 2. Every English side must exist in ESV.
    {
        let empty = HashSet::new();
        let have = present.get("ESV").unwrap_or(&empty);
        let mut missing: Vec<&str> = rows
            .iter()
            .map(|r| r.to_ref.as_str())
            .filter(|r| !have.contains(*r))
            .collect::<HashSet<_>>()
            .into_iter()
            .collect();
        missing.sort_unstable();
        if !missing.is_empty() {
            let head: Vec<&str> = missing.iter().take(8).copied().collect();
            problems.push(format!(
                "{} mapped English references absent from ESV: {}",
                missing.len(),
                py_list(&head)
            ));
        }
    }

    // 3. Repairing book identity must surface exactly the books VerseMap
    //    knows about: chapters whose verse counts diverge, both sides.
    let mapped_books: HashSet<&str> = rows
        .iter()
        .map(|r| r.from_ref.split('.').next().unwrap_or(""))
        .collect();
    let mut heb_by_chapter: HashMap<(&str, u64), usize> = HashMap::new();
    let mut eng_by_chapter: HashMap<(&str, u64), usize> = HashMap::new();
    let empty = HashSet::new();
    for reference in present.get("LHB").unwrap_or(&empty) {
        let mut parts = reference.split('.');
        if let (Some(book), Some(chapter)) = (parts.next(), parts.next()) {
            if let Ok(chapter) = chapter.parse::<u64>() {
                *heb_by_chapter.entry((book, chapter)).or_default() += 1;
            }
        }
    }
    for reference in present.get("ESV").unwrap_or(&empty) {
        let mut parts = reference.split('.');
        if let (Some(book), Some(chapter)) = (parts.next(), parts.next()) {
            if let Ok(chapter) = chapter.parse::<u64>() {
                *eng_by_chapter.entry((book, chapter)).or_default() += 1;
            }
        }
    }
    let mut divergent: HashSet<&str> = HashSet::new();
    for (key, count) in &heb_by_chapter {
        if eng_by_chapter.get(key) != Some(count) {
            divergent.insert(key.0);
        }
    }
    if divergent != mapped_books {
        let mut left: Vec<&str> = divergent.iter().copied().collect();
        let mut right: Vec<&str> = mapped_books.iter().copied().collect();
        left.sort_unstable();
        right.sort_unstable();
        problems.push(format!(
            "divergent books {} != VerseMap books {}",
            py_list(&left),
            py_list(&right)
        ));
    }

    problems
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let get = |name: &str| {
        args.iter()
            .position(|a| a == name)
            .and_then(|i| args.get(i + 1).cloned())
            .unwrap_or_else(|| {
                eprintln!("{name} is required");
                std::process::exit(2);
            })
    };
    let books = edition_book_rows();
    if args.iter().any(|a| a == "--dump-books") {
        // Machine modes print only JSON.
        let rows: Vec<Value> = books
            .iter()
            .map(|b| {
                json!({
                    "edition_key": b.edition_key,
                    "code": b.code,
                    "osis_id": b.osis_id,
                    "ordinal": b.ordinal,
                })
            })
            .collect();
        println!("{}", Value::Array(rows));
        return;
    }
    let raw = std::fs::read(get("--verse-map")).unwrap();
    let source = args
        .iter()
        .position(|a| a == "--source")
        .and_then(|i| args.get(i + 1).cloned())
        .unwrap_or_else(|| "test".to_owned());
    let mappings = parse_verse_map(&raw, &source).unwrap_or_else(|err| {
        eprintln!("{err}");
        std::process::exit(1);
    });
    if args.iter().any(|a| a == "--dump-rows") {
        // Machine mode prints only JSON.
        let rows: Vec<Value> = mappings
            .iter()
            .map(|m| {
                json!({
                    "from_scheme": "hebrew",
                    "to_scheme": "english",
                    "from_ref": m.from_ref,
                    "to_ref": m.to_ref,
                    "from_part": m.from_part,
                    "to_part": m.to_part,
                    "mapping_type": m.mapping_type,
                    "source": m.source,
                })
            })
            .collect();
        println!("{}", Value::Array(rows));
        return;
    }
    let partials = mappings
        .iter()
        .filter(|m| m.mapping_type == "partial")
        .count();
    println!(
        "{} book codes, {} verse mappings from {source}",
        books.len(),
        mappings.len()
    );
    println!(
        "  {} full, {partials} partial, {} books",
        mappings.len() - partials,
        mappings
            .iter()
            .map(|m| m.from_ref.split('.').next().unwrap_or(""))
            .collect::<HashSet<_>>()
            .len()
    );
    let problems = check_map_shape(&mappings);
    if !problems.is_empty() {
        println!("REFUSED — the map does not validate:");
        for problem in problems.iter().take(10) {
            println!("  {problem}");
        }
        std::process::exit(1);
    }
    if let Some(i) = args.iter().position(|a| a == "--verses") {
        let verses_raw = std::fs::read_to_string(&args[i + 1]).unwrap();
        let verses: HashMap<String, Vec<(String, u64, u64)>> =
            serde_json::from_str(&verses_raw).unwrap();
        let corpus_problems = validate_against_corpus(&mappings, &books, &verses);
        if !corpus_problems.is_empty() {
            println!("REFUSED — the map does not agree with the corpus:");
            for problem in corpus_problems.iter().take(10) {
                println!("  {problem}");
            }
            std::process::exit(1);
        }
        println!("validated: shape and corpus agreement both clean");
    } else if problems.is_empty() {
        println!("validated: shape clean");
    }
}
